docs/etcd.md

```diff
@@ -0,0 +1,405 @@
+- `PrefixT[T]` and `KeyT[T]` generic types - abstractions for etcd prefix and key.
+  - `T` is the type of the entity, see [Schema](#schema).
+- A high-level wrapper for basic etcd operations, see [Basic Operations](#basic-operations).
+- Combination of basic operations into a transaction, see [TxnOp](#txnop).
+- Combination of basic operations and transactions into an atomic operation with multiple read phases, see [AtomicOp](#atomicop).
+- Connection of a processor/callback to each operation and its parts.
+- [Iterator](#iterator) loads all pages from the same revision snapshot.
+- High-level Watch API streams - with manual restarts and automatic restarts on network failure, see [Watch Streams](#watch-streams).
+- High-level sessions/leases - with retries on network failure, see [Session](#session).
+
+### Packages
+
+The `etcdop` package consists of the following sub-packages:
+- [etcdop](../internal/pkg/service/common/etcdop) - prefix, key, watch stream.
+- [etcdop/serde](../internal/pkg/service/common/etcdop/serde), see [Serde](#serde).
+- [etcdop/op](../internal/pkg/service/common/etcdop/op) - basic operations, [TxnOp](#txnop), [AtomicOp](#atomicop).
+- [etcdop/iterator](../internal/pkg/service/common/etcdop/iterator), see [Iterator](#iterator).
+
+**In the future, it would be beneficial to combine these sub-packages into a single `etcdop` package, possibly excluding `serde`.**
+
+### Serde
+
+The `serde` package provides encoding, decoding, and validation operations for any value stored in etcd.
+
+Entities are validated upon save, but also upon load.
+
+The `serde.Serde` is an argument of the `etcdop.NewTypedKey` and `etcdop.NewTypedPrefix` constructors.
+
+Supported encodings:
+- Currently, only the `JSON` encoding is implemented.
+- If necessary in the future, it is possible to implement additional encodings.
+  - For example, for performance or database size optimization.
+- Possible extensions could include `gzip` compression or binary [Protocol Buffers](https://protobuf.dev) encoding.
+- It is possible to work with several encodings simultaneously:
+  - This allows for maintaining backward compatibility.
+  - Or implementing gradual migration.
+  - For example, a header/prefix can be used to determine encoding:
+    - `{` or `[` for JSON
+    - `PB:` for Protocol Buffers
+    - `GZIP:` for compression.
+  - The disadvantage is more complicated debugging, as values will not be directly readable.
+
+### Schema
+
+When using the `etcdop` framework, we must start with a schema.
+
+The schema defines:
+- Prefixes:
+  - `etcdop.Prefix` type is for raw values, represented by bytes.
+  - `etcdop.PrefixT[T]` type is for values encoded/decoded by the [serde](#serde).
+- Keys:
+  - `etcdop.Key` type is for raw values, represented by bytes.
+  - `etcdop.KeyT[T]` type is for values encoded/decoded by the [serde](#serde).
+
+Example of defining a typed prefix and key:
+```go
+s := serde.NewJSON(validator.Validate)
```

Comment:
Is there any specific thing about `validator.Validate`? Should we always use this setup of validator or can this be omitted?

Reply:
I will add a link to the validator package.